tools/gen/impl_template

```diff
@@ -305,6 +305,39 @@ fail:
 	fromwire_fail(cursor, max);
 	return false;
 }
+bool ${tlv.name}_is_valid(const struct ${tlv.struct_name()} *record)
+{
+	size_t numfields = tal_count(record->fields);
+	bool first = true;
+	u64 prev_type = 0;
+	for (size_t i=0; i<numfields; i++) {
+		struct tlv_field *f = &record->fields[i];
+		if (f->numtype % 2 == 0 && f->meta == 0) {
+			/* BOLT #1:
+			 * - otherwise, if `type` is unknown:
+			 *   - if `type` is even:
+			 *     - MUST fail to parse the `tlv_stream`.
+			 *   - otherwise, if `type` is odd:
+			 *     - MUST discard the next `length` bytes.
+			 */
+			SUPERVERBOSE("Unknown even type in TLV");
+			return false;
+		} else if (!first && f->numtype <= prev_type) {
+			/* BOLT #1:
+			 *  - if decoded `type`s are not monotonically-increasing:
+			 *    - MUST fail to parse the `tlv_stream`.
+			 */
+			if (f->numtype == prev_type)
+				SUPERVERBOSE("duplicate tlv type");
+			else
+				SUPERVERBOSE("invalid ordering");
+			return false;
+		}
+		first = false;
```

Comment:
This code doesn't actually work; we don't update prev_type!

We had tests for the generic code which I think we should replicate here which should have caught this?

Reply:
I'll migrate the `run-tlvstream` tests to use the typesafe wrapper and remove the old ones. :+1: